### PR TITLE
Add tests for Testimonial component

### PR DESCRIPTION
**Pull request by joshfarrant**
## Summary

Backfills tests for Testimonial component to bring [coverage up to 100%](https://github.com/primer/brand/actions/runs/16293248338/job/46008099945?pr=1107#step:10:227).

## Supporting resources (related issues, external links, etc)

- Towards https://github.com/github/primer/issues/5412

## Contributor checklist

- [x] All new and existing CI checks pass
- [x] Tests prove that the feature works and covers both happy and unhappy paths
- [x] Any drop in coverage, breaking changes or regressions have been documented above
- [x] UI Changes contain new visual snapshots (generated by adding `update snapshots` label to the PR)
- [x] All developer debugging and non-functional logging has been removed
- [x] Related issues have been referenced in the PR description

## Reviewer checklist

- [ ] Check that pull request and proposed changes adhere to our [contribution guidelines](../../CONTRIBUTING.md) and [code of conduct](../../CODE_OF_CONDUCT.md)
- [ ] Check that tests prove the feature works and covers both happy and unhappy paths
- [ ] Check that there aren't other open Pull Requests for the same update/change

### CODE REVIEW

File: packages/react/src/Testimonial/Testimonial.test.tsx [hunk 1]
```diff
@@ -0,0 +1,516 @@
+import React from 'react'
+import {render} from '@testing-library/react'
+import '@testing-library/jest-dom'
+import {axe, toHaveNoViolations} from 'jest-axe'
+import {Testimonial} from './Testimonial'
+
+expect.extend(toHaveNoViolations)
+
+const leftDoubleQuote = '“'
```
Comment (rezrah) [inline packages/react/src/Testimonial/Testimonial.test.tsx]: Could we move this inside the describe block with the other mock values?

File: packages/react/src/Testimonial/Testimonial.test.tsx [hunk 2]
```diff
@@ -0,0 +1,516 @@
...
+  })
+
+  it('renders quote mark with default color', () => {
+    const {getByText} = render(<DefaultTestimonial />)
+
+    const quoteMark = getByText(leftDoubleQuote)
+    expect(quoteMark).toHaveClass('Testimonial__quoteMark--default')
+  })
+
+  it('applies custom className', () => {
```
Comment (rezrah) [inline packages/react/src/Testimonial/Testimonial.test.tsx]:
```suggestion
  it('allows forwarding of custom classes', () => {
```

File: packages/react/src/Testimonial/Testimonial.test.tsx [hunk 3]
```diff
@@ -0,0 +1,516 @@
...
+    const {getByRole} = render(
+      <Testimonial className="custom-testimonial">
+        <Testimonial.Quote>Quote text</Testimonial.Quote>
+        <Testimonial.Name>Name</Testimonial.Name>
+      </Testimonial>,
+    )
+
+    expect(getByRole('figure')).toHaveClass('custom-testimonial')
```
Comment (rezrah) [inline packages/react/src/Testimonial/Testimonial.test.tsx]: Could we also ensure custom classes don't overwrite the default classes please? One of the general edge case tests that should apply to all components, as it's led to bug reports in the past.

Example:

```suggestion
    const el = getByRole('figure');
    expect(el).toHaveClass('custom-testimonial')
    expect(el).toHaveClass('Testimonial')'
```

Also just to note that for readability and style preference, can we extract the element matcher outside of assertion please

File: packages/react/src/Testimonial/Testimonial.test.tsx [hunk 4]
```diff
@@ -0,0 +1,516 @@
...
+  })
+
+  it('supports RefObject for Testimonial component', () => {
+    const refObject = React.createRef<HTMLElement>()
```
Comment (rezrah) [inline packages/react/src/Testimonial/Testimonial.test.tsx]: Can we prefer useRef instead of createRef for these tests to be more representative of end-user usage?

File: packages/react/src/Testimonial/Testimonial.test.tsx [hunk 5]
```diff
@@ -0,0 +1,516 @@
...
+      <Testimonial className="custom-testimonial">
+        <Testimonial.Quote>Quote text</Testimonial.Quote>
+        <Testimonial.Name>Name</Testimonial.Name>
+      </Testimonial>,
+    )
+
+    expect(getByRole('figure')).toHaveClass('custom-testimonial')
+  })
+
+  it('supports RefObject for Testimonial component', () => {
```
Comment (rezrah) [inline packages/react/src/Testimonial/Testimonial.test.tsx]: For conciceness, and could we remove from other case names too please.

```suggestion
  it('supports RefObject refs', () => {
```

File: packages/react/src/Testimonial/Testimonial.test.tsx [hunk 6]
```diff
@@ -0,0 +1,516 @@
...
+  it('applies custom quote mark color', () => {
+    const {getByText} = render(
+      <Testimonial quoteMarkColor="blue">
+        <Testimonial.Quote>Quote text</Testimonial.Quote>
+        <Testimonial.Name>Name</Testimonial.Name>
+      </Testimonial>,
+    )
+
+    const quoteMark = getByText(leftDoubleQuote)
+    expect(quoteMark).toHaveClass('Testimonial__quoteMark--blue')
```
Comment (rezrah) [inline packages/react/src/Testimonial/Testimonial.test.tsx]: In some other tests, I like to put `expectedColor` at the top to set expectations early. Can we do the same here please. 

E.g. 

```suggestion
    const expectedColor = 'blue'
    
    const {getByText} = render(
      <Testimonial quoteMarkColor={expectedColor}>
        <Testimonial.Quote>Quote text</Testimonial.Quote>
        <Testimonial.Name>Name</Testimonial.Name>
      </Testimonial>,
    )

    const quoteMark = getByText(leftDoubleQuote)
    expect(quoteMark).toHaveClass(`Testimonial__quoteMark--${expectedColor}`)

```

File: packages/react/src/Testimonial/Testimonial.test.tsx [hunk 7]
```diff
@@ -0,0 +1,516 @@
...
+        </Testimonial.Quote>
+        <Testimonial.Name>Name</Testimonial.Name>
+      </Testimonial>,
+    )
+
+    const quoteText = getByText('This is', {exact: false})
+    expect(quoteText).toHaveClass('Testimonial-quote--muted')
+  })
+
+  it('applies muted variant when quote contains emphasized text', () => {
```
Comment (rezrah) [inline packages/react/src/Testimonial/Testimonial.test.tsx]:
```suggestion
  it('applies muted text color when the quote contains emphasized text', () => {
```

File: packages/react/src/Testimonial/Testimonial.test.tsx [hunk 8]
```diff
@@ -0,0 +1,516 @@
...
+        <Testimonial.Avatar src={mockAvatarSrc} alt={mockAvatarAlt} />
+      </Testimonial>,
+    )
+
+    const avatar = getByRole('img', {name: mockAvatarAlt})
+    expect(avatar).toBeInTheDocument()
+    expect(avatar).toHaveAttribute('src', mockAvatarSrc)
+  })
+
+  it('applies testimonial avatar class', () => {
```
Comment (rezrah) [inline packages/react/src/Testimonial/Testimonial.test.tsx]: What is this testing for? It feels like it could be rolled into the tests above? Could consolidate all individual tests for avater into one case?

File: packages/react/src/Testimonial/Testimonial.test.tsx [hunk 9]
```diff
@@ -0,0 +1,516 @@
...
+    const name = getByText(mockName)
+
+    const figureChildren = Array.from(figure.children)
+    const quoteIndex = figureChildren.findIndex(child => child.contains(quote))
+    const nameIndex = figureChildren.findIndex(child => child.contains(name))
+
+    expect(quoteIndex).toBeLessThan(nameIndex)
+  })
+
+  it('renders avatar, logo, and name in media section', () => {
```
Comment (rezrah) [inline packages/react/src/Testimonial/Testimonial.test.tsx]: This one feels a little like implementation detail. It seems like we're expecting elements to render in a specific, generic div but not sure if that's necessary. wdyt?

File: packages/react/src/Testimonial/Testimonial.test.tsx [hunk 10]
```diff
@@ -0,0 +1,516 @@
...
+        <Testimonial.Name>Name</Testimonial.Name>
+      </Testimonial>,
+    )
+
+    expect(getByRole('figure')).toBeInTheDocument()
+    expect(getByText('Quote text')).toBeInTheDocument()
+    expect(getByText('Name')).toBeInTheDocument()
+  })
+
+  it('handles mixed children types', () => {
```
Comment (rezrah) [inline packages/react/src/Testimonial/Testimonial.test.tsx]:
```suggestion
  it('filters out children that aren't part of the Testimonial namespace', () => {
```

File: packages/react/src/Testimonial/Testimonial.test.tsx (hunk 5 above)
Comment (joshfarrant) [reply packages/react/src/Testimonial/Testimonial.test.tsx]: I can do, but here I'm specifically testing the Testimonial component (as opposed to one of the compound components), so I'd lean towards being explicit here. For example, we have all of these tests within this test suite

```
it('supports functional ref for Testimonial component')
it('supports functional ref for Testimonial.Name component')
it('supports functional ref for Testimonial.Logo component')
it('supports functional ref for Testimonial.Quote component')
```

Happy to change if you'd like, although this could be a good use-case for nested `describe` blocks so we can keep test names clean whilst still clearly scoping them to their related component.

File: packages/react/src/Testimonial/Testimonial.test.tsx (hunk 4 above)
Comment (joshfarrant) [reply packages/react/src/Testimonial/Testimonial.test.tsx]: I went with `createRef` here for a couple of reasons.

1. To keep things consistent with [how we're testing in the Checkbox component](https://github.com/primer/brand/blob/main/packages/react/src/forms/Checkbox/Checkbox.test.tsx#L90)
2. Testing with `useRef` is a bit awkward as `useRef` is a hook, so will need to be called from within a component. The following would work, but it doesn't feel great IMO
```tsx
  it('supports RefObject for Testimonial component', () => {
    let refObject

    const TestComponent = () => {
      refObject = React.useRef<HTMLElement>(null)

      return (
        <Testimonial ref={refObject}>
          <Testimonial.Quote>Quote text</Testimonial.Quote>
          <Testimonial.Name>Name</Testimonial.Name>
        </Testimonial>
      )
    }

    const {getByRole} = render(<TestComponent />)
    const figure = getByRole('figure')
    expect(refObject.current).toBe(figure)
  })
```

Both `useRef()` and `React.createRef()` return the same underlying `RefObject` as far as I'm aware.

If you'd like me to implement the above approach though just let me know and I'm happy to do so.

File: packages/react/src/Testimonial/Testimonial.test.tsx (hunk 9 above)
Comment (joshfarrant) [reply packages/react/src/Testimonial/Testimonial.test.tsx]: Yep happy to remove it 👍

File: packages/react/src/Testimonial/Testimonial.test.tsx (hunk 8 above)
Comment (joshfarrant) [reply packages/react/src/Testimonial/Testimonial.test.tsx]: Looks like something I was playing around with then didn't remove, sorry about that, removed

File: packages/react/src/Testimonial/Testimonial.test.tsx (hunk 6 above)
Comment (joshfarrant) [reply packages/react/src/Testimonial/Testimonial.test.tsx]: I tend to favour the explicit approach for simple tests like this. It's immediately clear we're passing 'blue' and expecting the 'blue' class — we don't have to deal with mentally tracking a variable or building template strings.

Happy to do it if you strongly prefer this approach, but adding unnecessary logic and abstraction to a test feels like it goes against good testing practise to me.

File: packages/react/src/Testimonial/Testimonial.test.tsx (hunk 5 above)
Comment (rezrah) [reply packages/react/src/Testimonial/Testimonial.test.tsx]: ah cool, makes sense. Happy to keep it this way, but could we please update the first one to be: `it supports functional ref for the base|root (whichever ones makes sense) Testimonial component`. IMO this sets an earlier expectaion that others similar tests will follow for the child components.

File: packages/react/src/Testimonial/Testimonial.test.tsx (hunk 4 above)
Comment (rezrah) [reply packages/react/src/Testimonial/Testimonial.test.tsx]: Yeah, makes sense. I was wondering if we could do what we did for SubNav [here](https://github.com/primer/brand/blob/main/packages/react/src/SubNav/SubNav.test.tsx#L274).

Comment (rezrah) [reply packages/react/src/Testimonial/Testimonial.test.tsx]: The reason being that I think it mimics how users will use the component. 1) Creating the wrapper 2) Create a ref with the hook 3) pass to our component 4) Doing something with it in their own handler or effect.

File: packages/react/src/Testimonial/Testimonial.test.tsx (hunk 6 above)
Comment (rezrah) [reply packages/react/src/Testimonial/Testimonial.test.tsx]: Sounds good, fine to keep it this way. Might come down to a style preference, and I don't have a super strong opinion on it right now. I think in many other parts of the codebase we do what I was suggesting. Could you clarify the part of "goes against good testing practise to me." please, as this feels more like a style preference to me.

File: packages/react/src/Testimonial/Testimonial.test.tsx [hunk 11]
```diff
@@ -0,0 +1,515 @@
...
+        <Testimonial.Quote>Quote text</Testimonial.Quote>
+        <Testimonial.Name>Name</Testimonial.Name>
+      </Testimonial>,
+    )
+
+    const quoteMark = getByText(leftDoubleQuote)
+    expect(quoteMark).toHaveClass('Testimonial__quoteMark--blue')
+  })
+
+  it('applies custom color as CSS variable', () => {
```
Comment (rezrah) [inline packages/react/src/Testimonial/Testimonial.test.tsx]: Is this test case different from the one above? Feels like they are part of the same logic, as they share the same color prop. Could we consider merging them into one test?

File: packages/react/src/Testimonial/Testimonial.test.tsx (hunk 6 above)
Comment (joshfarrant) [reply packages/react/src/Testimonial/Testimonial.test.tsx]: > Could you clarify the part of "goes against good testing practise to me." please

Here I'm referring to the practise of having tests contain minimal logic and be "obviously correct" at a glance. Adding variables and template literals for single-use values introduces small amounts of logic that require mental parsing, whereas hardcoded values are immediately clear.

As I say, if you did feel strongly that we should continue with your suggested approach then we can do though.

File: packages/react/src/Testimonial/Testimonial.test.tsx (hunk 11 above)
Comment (joshfarrant) [reply packages/react/src/Testimonial/Testimonial.test.tsx]: Yes good spot, thanks 👍

File: packages/react/src/Testimonial/Testimonial.test.tsx [hunk 12]
```diff
@@ -0,0 +1,515 @@
...
+      <Testimonial>
+        <Testimonial.Quote>Quote text</Testimonial.Quote>
+        <Testimonial.Name>Name</Testimonial.Name>
+        <Testimonial.Avatar src={mockAvatarSrc} alt={mockAvatarAlt} size={64} />
+      </Testimonial>,
+    )
+
+    const avatarContainer = getByAltText(mockAvatarAlt)
+    expect(avatarContainer).toBeInTheDocument()
+  })
```
Comment (rezrah) [inline packages/react/src/Testimonial/Testimonial.test.tsx]: I think the test name here is wrong? Prop value is `64` but test name is `48`?. I might be missing something? 🤔. Should this be asserting for a class or specific dimension property too rather than checking it's in the document?

Comment (joshfarrant) [reply packages/react/src/Testimonial/Testimonial.test.tsx]: The test name could be clearer here. I suppose what I'm actually testing is that the size of the `Avatar` _can't_ be overridden as it's intentionally fixed to 48px. Let me make that a bit clearer, and improve the assertion.

Comment (joshfarrant) [reply packages/react/src/Testimonial/Testimonial.test.tsx]: Improved in 412e52de200b7633f1db7f535451c46bcde24a6a, lmk what you think

File: packages/react/src/Testimonial/Testimonial.test.tsx (hunk 6 above)
Comment (rezrah) [reply packages/react/src/Testimonial/Testimonial.test.tsx]: Cool, thanks for clarifying. No strong opinions on this beyond the mentioned inconsistency in other test files. Ironically I see it the other, and find it harder to read this way, because I don't really care what the entire className was, just that the expectedValue appears in the final class. Fine to leave as-is for this PR though.

File: packages/react/src/Testimonial/Testimonial.test.tsx (hunk 12 above)
Comment (rezrah) [reply packages/react/src/Testimonial/Testimonial.test.tsx]: Yep, makes sense 🙌

**Review by rezrah — APPROVED** [review]
